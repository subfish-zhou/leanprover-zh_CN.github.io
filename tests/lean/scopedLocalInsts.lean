structure A where
  x : Nat
  y : Nat

namespace Foo

scoped instance : ToString A where
  toString a := s!"A.mk {a.x} {a.y}"

end Foo

#eval toString { x := 10, y := 20 : A } -- Error: failed to synthesize `ToString A`

section Sec1
open Foo

#eval toString { x := 10, y := 20 : A } -- works

end Sec1 -- scoped instance `ToString A` is not available anymore

#eval toString { x := 10, y := 20 : A } -- Error

section Sec2

local instance : ToString A where
  toString a := s!"\{ x := {a.x}, y := {a.y} }"

#eval toString { x := 10, y := 20 : A } -- works, using local instance

end Sec2 -- local instance `ToString A` is not available anymore

#eval toString { x := 10, y := 20 : A } -- Error

open Foo

#eval toString { x := 10, y := 20 : A } -- works
