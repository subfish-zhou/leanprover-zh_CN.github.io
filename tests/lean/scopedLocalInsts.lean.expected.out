scopedLocalInsts.lean:12:6-12:39: error: failed to synthesize instance
  ToString A
scopedLocalInsts.lean:12:0-12:39: error: failed to synthesize
  Lean.Eval ?m
"A.mk 10 20"
scopedLocalInsts.lean:21:6-21:39: error: failed to synthesize instance
  ToString A
scopedLocalInsts.lean:21:0-21:39: error: failed to synthesize
  Lean.Eval ?m
"{ x := 10, y := 20 }"
scopedLocalInsts.lean:32:6-32:39: error: failed to synthesize instance
  ToString A
scopedLocalInsts.lean:32:0-32:39: error: failed to synthesize
  Lean.Eval ?m
"A.mk 10 20"
